Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -45,6 +52,173 @@ namespace WebKit {
+                int fd = *static_cast<int*>(userData);
+                while (length) {
+                    auto written = write(fd, data, length);
+                    if (written == -1)
+                        return CAIRO_STATUS_WRITE_ERROR;
+
+                    length -= written;
+                    data += written;
+                }
+                return CAIRO_STATUS_SUCCESS;
+            }, &fd);
+
+            close(fd);
+
+            return status == CAIRO_STATUS_SUCCESS ? filename.get() : CString();
+        };
+
+        auto addResult = m_iconCache.add(iconURL, std::pair<uint32_t, CString>({ 0, CString() }));
+        if (addResult.isNewEntry) {
+            auto path = writeIconToTemporaryFile(icon);
+            if (path.isNull()) {
+                m_iconCache.remove(addResult.iterator);
+                return nullptr;
+            }
+            addResult.iterator->value = { 1, WTFMove(path) };
+        } else
+            addResult.iterator->value.first++;
+
+        return std::get<CString>(addResult.iterator->value.second).data();
+    }
+
+    GBytes* iconBytes(const String& iconURL, const RefPtr<WebCore::Image>& icon)
+    {
+        if (!icon)
+            return nullptr;
+
+        auto writeIconToBuffer = [](const RefPtr<WebCore::Image>& icon) -> GRefPtr<GBytes> {
+            auto nativeImage = icon->nativeImage();
+            if (!nativeImage)
+                return nullptr;
+
+            auto surface = nativeImage->platformImage();
+            if (!surface)
+                return nullptr;
+
+            GRefPtr<GByteArray> buffer = adoptGRef(g_byte_array_new());
+            auto status = cairo_surface_write_to_png_stream(surface.get(), [](void* userData, const unsigned char* data, unsigned length) -> cairo_status_t {
+                auto* buffer = static_cast<GByteArray*>(userData);
+                g_byte_array_append(buffer, data, length);
+                return CAIRO_STATUS_SUCCESS;
+            }, buffer.get());
+
+            if (status != CAIRO_STATUS_SUCCESS)
+                return nullptr;
+
+            return adoptGRef(g_byte_array_free_to_bytes(buffer.leakRef()));
+        };
+
+        auto addResult = m_iconCache.add(iconURL, std::pair<uint32_t, GRefPtr<GBytes>>({ 0, nullptr }));
+        if (addResult.isNewEntry) {
+            auto bytes = writeIconToBuffer(icon);
+            if (!bytes) {
+                m_iconCache.remove(addResult.iterator);
+                return nullptr;
+            }
+            addResult.iterator->value = { 1, WTFMove(bytes) };
+        } else
+            addResult.iterator->value.first++;
+
+        return std::get<GRefPtr<GBytes>>(addResult.iterator->value.second).get();
+    }
+
+    void unuseIcon(const String& iconURL)
+    {
+        auto it = m_iconCache.find(iconURL);
+        if (it == m_iconCache.end())
+            return;
+
+        if (!it->value.first)
+            return;
```

Comment:
In this case, deletion is already pending, right? It took me too long to figure this out, so it's probably worth adding a quick comment to explain.

Reply:
No, this means it's already deleted (or scheduled) at this point, I'm not sure it can really happen, maybe it should be an assert, but it's to ensure we don't overflow the unsigned.